### PR TITLE
Make logs in invokeExceptionCaught consistent and full

### PR DESCRIPTION
**Pull request by Spikhalskiy**
Motivation:

In case of exception in invokeExceptionCaught() only original exception passed to invokeExceptionCaught() will be logged on any log level.
AbstractChannelHandlerContext and CombinedChannelDuplexHandler log different exceptions.

Modifications:

Fix inconsistent logging code and add ability to see both stacktraces on DEBUG level.

Result:

Both handlers log now both original exception and thrown from invokeExceptionCaught. To see full stacktrace of exception thrown from invokeExceptionCaught DEBUG log level must be enabled.

Could be applied to 4.0 branch too.

### CODE REVIEW

**Comment by normanmaurer** [comment]
Let me check

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 1]
```diff
@@ -276,11 +276,11 @@ private void invokeExceptionCaught(final Throwable cause) {
         if (isAdded()) {
             try {
                 handler().exceptionCaught(this, cause);
-            } catch (Throwable t) {
+            } catch (Throwable error) {
                 if (logger.isWarnEnabled()) {
                     logger.warn(
-                            "An exception was thrown by a user handler's " +
-                                    "exceptionCaught() method while handling the following exception:", cause);
+                        "An exception '{}' was thrown by a user handler's exceptionCaught() " +
```
Comment (johnou) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Not sure logging exception.tostring here is going to help much,   the broken exceptionCaught implementation will still need to be fixed without knowing the line number.

Comment (Spikhalskiy) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Agree. It's just a solution that is slightly better without overhead in terms of huge logs growing as a result of output of two stacktraces in one message. So, if it's fine - could be reworked to full stacktraces.

Comment (johnou) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: maybe at debug level?

Comment (Spikhalskiy) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Yep, may be good to print current version on WARN and full version with two stack traces on DEBUG level. @normanmaurer WDYT?

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: @Spikhalskiy +1

**Comment by normanmaurer** [comment]
@Spikhalskiy please make the change and also update the commit message to match our template:

http://netty.io/wiki/writing-a-commit-message.html

**Comment by normanmaurer** [comment]
@Spikhalskiy just ping me once you are done and ready for another review

**Comment by Spikhalskiy** [comment]
@normanmaurer fixed, please check when you get a chance.

File: common/src/main/java/io/netty/util/internal/ExceptionUtils.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2012 The Netty Project
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/ExceptionUtils.java]: 2016

File: common/src/main/java/io/netty/util/internal/ExceptionUtils.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+public final class ExceptionUtils {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/ExceptionUtils.java]: Call it `ThrowableUtils`.

File: common/src/main/java/io/netty/util/internal/ExceptionUtils.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    /**
+     * <p>Gets the stack trace from a Throwable as a String.</p>
+     *
+     * This method has been cherry-picked from {@link org.apache.commons.lang3.exception.ExceptionUtils}
+     *
+     * @param throwable  the <code>Throwable</code> to be examined
+     * @return the stack trace as generated by the exception's
+     *  <code>printStackTrace(PrintWriter)</code> method
+     */
+    public static String getStackTrace(final Throwable throwable) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/ExceptionUtils.java]: Please replace this method with this that can be found here:
https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L700

And also reuse it there then :) 

With this you can also remove the link to lang3.

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 5]
```diff
@@ -274,11 +275,18 @@ private void invokeExceptionCaught(final Throwable cause) {
         if (isAdded()) {
             try {
                 handler().exceptionCaught(this, cause);
-            } catch (Throwable t) {
-                if (logger.isWarnEnabled()) {
+            } catch (Throwable error) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug(
+                        "An exception\n{}" +
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Also can you use the same "formatting" as we do here:
https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L694

File: transport/src/main/java/io/netty/channel/CombinedChannelDuplexHandler.java [hunk 6]
```diff
@@ -142,10 +143,17 @@ public ChannelHandlerContext fireExceptionCaught(Throwable cause) {
                         // as well
                         outboundHandler.exceptionCaught(outboundCtx, cause);
                     } catch (Throwable error) {
-                        if (logger.isWarnEnabled()) {
+                        if (logger.isDebugEnabled()) {
+                            logger.debug(
+                                    "An exception\n{}" +
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/CombinedChannelDuplexHandler.java]: same as above

**Comment by normanmaurer** [comment]
@Spikhalskiy sorry a few more suggestions.

File: common/src/main/java/io/netty/util/internal/ExceptionUtils.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        final StringWriter sw = new StringWriter();
+        final PrintWriter pw = new PrintWriter(sw, true);
+        throwable.printStackTrace(pw);
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/ExceptionUtils.java]: CRITICAL Use a logger to log this exception. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1148)

**Comment by Spikhalskiy** [comment]
@Done, netkins comment could be ignored I think

**Comment by normanmaurer** [comment]
Yes

**Comment by normanmaurer** [comment]
@Spikhalskiy thanks!

Cherry-picked into 4.0 (a2ce9f2cb424f92b0587944391ca099630cad965) and 4.1 (428c61673bf5acd4f4807d950bb03e13d83f2223)

File: common/src/main/java/io/netty/util/internal/ThrowableUtils.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+     * <p>Gets the stack trace from a Throwable as a String.</p>
+     *
+     * @param throwable the <code>Throwable</code> to be examined
+     * @return the stack trace as generated by the exception's
+     *  <code>printStackTrace(PrintWriter)</code> method
+     */
+    public static String stackTraceToString(Throwable cause) {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        PrintStream pout = new PrintStream(out);
+        cause.printStackTrace(pout);
```
Comment (netkins) [inline common/src/main/java/io/netty/util/internal/ThrowableUtils.java]: CRITICAL Use a logger to log this exception. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1148)

**Comment by Spikhalskiy** [comment]
@normanmaurer Looks like you didn't cherry-pick to 4.0 or forgot to push branch.

**Comment by normanmaurer** [comment]
Let me check

**Pull request closed without merging**